src/metabase/lib/util.cljc

```diff
@@ -325,45 +325,40 @@
            conjunction
            (last coll)))))))
 
-(defn update-stages-ignore-joins
-  "Like [[update-stages]], but does not recurse into the stages inside joins.
+(defn update-query-and-joins
+  "Update the top level of a query, and all of the joins (i.e., all the things that have `:stages`). Done in a
+  depth-first manner.
 
   `f` has the signature
 
-    (f query stage-number stage)"
+    (f query-or-join)"
   [query f]
-  (reduce
-   (fn [query stage-number]
-     (update-in query [:stages stage-number] (fn [stage]
-                                               (f query stage-number stage))))
-   query
-   (range 0 (count (:stages query)))))
+  (letfn [(update-join [join]
+            (update-query-and-joins join f))
+          (update-joins [joins]
+            (mapv update-join joins))
+          (update-stage [stage]
+            (m/update-existing stage :joins update-joins))
+          (update-stages [stages]
+            (mapv update-stage stages))]
+    (-> query
+        (update :stages update-stages)
+        f)))
 
 (defn update-stages
-  "Apply function `f` to every stage of a query, depth-first. Also applied to all query stages.
+  "Apply function `f` to every stage of a query, depth-first. Also applied to all join stages.
 
   `f` has the signature
 
-    (f query stage-number stage)
-
-  `query` reflects the results of the previous call to `f`.
+    (f stage)
 
-  As a convenience, if `f` returns nil, the original stage will be used without changes."
+  As a convenience, if `f` returns `nil`, the original stage will be used without changes."
   [query f]
-  (letfn [(update-join [join]
-            (-> query
-                (assoc :stages (:stages join))
-                (update-stages f)
-                :stages))
-          (update-joins [joins]
-            (mapv update-join joins))]
-    (update-stages-ignore-joins
-     query
-     (fn [query stage-number stage]
-       (let [stage (cond-> stage
-                     (:joins stage)
-                     update-joins)]
-         (f query stage-number stage))))))
+  (letfn [(update-stages [stages]
+            (mapv #(or (f %) %) stages))
+          (update-top-level [top-level]
+            (update top-level :stages update-stages))]
+    (update-query-and-joins query update-top-level)))
```

Comment:
So the old `update-stages` turned out to be broken, but it was also unused. I fixed it; it is used in #31788, but I thought I might as well get the fixed version in sooner rather than later since that PR will probably take a while to land.